tracing: control net/trace and lightstep via cluster settings

RaduBerinde opened the pull request:
See individual commits. The TL;DR is that we now have two cluster settings that control net/trace and the lightstep token. The default is for them to be off in which case no tracing happens (other than snowball tracing when requested).

I plan to run some more manual tests but it seems to work so far.
CC @dt for the naming of the settings (is there a doc somewhere with guidelines?). Note that we may add more lightstep settings like `trace.lightstep.address` and whatnot.

RaduBerinde committed:
tracing: control net/trace via cluster setting

This change replaces `COCKROACH_ENABLE_TRACING` with a cluster setting
`trace.debug.enable`. The semantics are changed slightly: the setting controls
whether we post traces to the debug page (via net/trace) and is now independent
of `COCKROACH_LIGHTSTEP_TOKEN`. If the latter is specified, lightstep will be
used regardless of `trace.debug.enable`.

Updates #16165.

RaduBerinde committed:
tracing: control lightstep token via cluster setting

Replacing `COCKROACH_LIGHSTEP_TOKEN` env var with `trace.lightstep.token` cluster
setting. Now we can enable/disable lightstep without restarting. Note that this
setting is independent of other settings; if it is set, lightstep will be
enabled regardless of `trace.debug.enable`.

When enabling lightstep, existing span hierarchies won't use it - only new root
spans and their descendants.

If necessary, we can easily add a setting for collector address in the future.

Fixes #16165.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16265)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but if there were any complications related to switching the flags on and off in the middle of things, I might've missed them.

---

Reviewed 12 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/tracing/tracer.go, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ8Jwb1q-fPqIIoEZF:-KlZ8Jwb1q-fPqIIoEZG:b42ujgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3409859c25c6f47a9411d7c3bceeaddae4c2e2/pkg/util/tracing/tracer.go#L159)):*
> ```Go
> 	// is disabled: if we have no options or a single SpanReference (the common
> 	// case) with a noop context, return a noop span now.
> 	if len(opts) == 1 && t.lightstep == nil {
> ```

`len(opts)` is never `0`? And is it set in stone that if there's exactly one opt, that it's a SpanReference? Perhaps a comment would be helpful here.

---

*[pkg/util/tracing/tracer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ9-IxStLMQquZhp24:-KlZ9-IxStLMQquZhp25:b-dwhq8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf9ececbaa0d3d1bb9ce311ddea313a8239de2f3/pkg/util/tracing/tracer.go#L201)):*
> ```Go
> 	// is disabled: if we have no options or a single SpanReference (the common
> 	// case) with a noop context, return a noop span now.
> 	if len(opts) == 1 {
> ```

This change is necessary to make sure that we don't mix noop and non-noop, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-:-KlZ9T0ZFC_CJqboHD4Y:b-5apb5i)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Fantastic! This will be so much more pleasant for users, rather than having to restart nodes to enable tracing.

dt commented:
s/@dt/@knz/ for naming bikeshed, but yes, he did write a guide to naming settings https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/settings_table.md#how-to-name-cluster-settings

RaduBerinde commented:
@tschottdorf If we turn lightstep off, any subsequent spans will not use lightstep. If we turn it on, children of existing spans won't use it but new spans will.

If we turn it off and on (or in the future, if we reconfigure the collector address), there may be spans that use the old one at the same time there are spans using the new one (which should be fine).

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/tracing/tracer.go, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ8Jwb1q-fPqIIoEZF:-KlZEBRIiMxewEGZfGmy:b-s3culj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3409859c25c6f47a9411d7c3bceeaddae4c2e2/pkg/util/tracing/tracer.go#L159)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`len(opts)` is never `0`? And is it set in stone that if there's exactly one opt, that it's a SpanReference? Perhaps a comment would be helpful here.
</blockquote></details>

The `len(opts) == 0` case is moved below (we need to read the options for that case).

The comment above explains it - the common cases are to have no option or just a parent reference option. This is all just to avoid doing any allocations when tracing is disabled.

---

*[pkg/util/tracing/tracer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ9-IxStLMQquZhp24:-KlZDjzypv7g2a_WAvcm:b-4m9yt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf9ececbaa0d3d1bb9ce311ddea313a8239de2f3/pkg/util/tracing/tracer.go#L201)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This change is necessary to make sure that we don't mix noop and non-noop, right?
</blockquote></details>

It's not important for correctness, I just want to avoid reading the lightstep pointer in the "fast path".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-:-KlZD5SLxkh-RDBGd1BW:b-thkp1s)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 12 files at r1, 4 of 4 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/kv_test.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZ5SFgLrKGK2YJr9S:-KlZZ5SFgLrKGK2YJr9T:b-b4jaz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/kv_test.go#L44)):*
> ```Go
> 
> func newKVWriteBatch(b *testing.B) kvInterface {
> 	enableTracing := tracing.Disable()
> ```

don't we still want to disable tracing here? this is performance sensitive.

---

*[pkg/ccl/sqlccl/load_test.go, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZA1sICALzJHETRrU:-KlZZA1tBKFNHigmtn4v:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/load_test.go#L64)):*
> ```Go
> 	// documentation's description. We're getting useful information out of it,
> 	// but this is not a pattern to cargo-cult.
> 	defer tracing.Disable()()
> ```

ditto

---

*[pkg/sql/trace_test.go, line 101 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZPrB2lJ2Oo0rbqg0:-KlZZPrB2lJ2Oo0rbqg1:bwtvw81) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b05cd5f371567603d601589902128dd2c296bd5/pkg/sql/trace_test.go#L101)):*
> ```Go
> 
> 			if _, err := cluster.ServerConn(0).Exec(
> 				fmt.Sprintf(`SET CLUSTER SETTING trace.debug.enable = %t;`, enableTr),
> ```

nit: no need for the semicolon. 

no placeholders here, huh?

---

*[pkg/util/tracing/tracer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ9-IxStLMQquZhp24:-KlZZyhiM-IAFeVRb_QS:brn9yqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf9ececbaa0d3d1bb9ce311ddea313a8239de2f3/pkg/util/tracing/tracer.go#L201)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's not important for correctness, I just want to avoid reading the lightstep pointer in the "fast path".
</blockquote></details>

subjectively I find this stuff harder to read than the previous switch-case style.

---

*[pkg/util/tracing/tracer.go, line 80 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ_s4hUD3J9fFNVB0A:-KlZ_s4hUD3J9fFNVB0B:b-dod3ko) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L80)):*
> ```Go
> func updateLightstep() {
> 	token := lightstepToken.Get()
> 	if token == "" {
> ```

optional: inline the `:=` here

---

*[pkg/util/tracing/tracer.go, line 96 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZaAXqeP1RKK47g0Ti:-KlZaAXqeP1RKK47g0Tj:b-n1lnai) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L96)):*
> ```Go
> 
> func getLightstep() opentracing.Tracer {
> 	ptr := atomic.LoadPointer(&lightstepPtr)
> ```

golf nit: easier to read `if ptr := ...; ptr != nil { return ...}; return nil`

---

*[pkg/util/tracing/tracer.go, line 291 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZb4aqE0HOZr8uivrV:-KlZb4aqE0HOZr8uivrW:bg2krlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L291)):*
> ```Go
> 		}
> 		s.lightstep = lsTr.StartSpan(operationName, lsOpts...)
> 		s.TraceID, s.SpanID = getLightstepSpanIDs(s.lightstep.Tracer(), s.lightstep.Context())
> ```

what's the difference between `s.lightstep.Tracer()` and `lsTr`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/kv_test.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZ5SFgLrKGK2YJr9S:-KlZc7JmwrCdzx-5Q2I7:b7gxhav) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/kv_test.go#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

don't we still want to disable tracing here? this is performance sensitive.
</blockquote></details>

Tracing is now disabled by default and I doubt that will change. I kept the disablement in some tests where it was easy enough but I don't think it's necessary.

---

*[pkg/sql/trace_test.go, line 101 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZPrB2lJ2Oo0rbqg0:-KlZfVta1_P_D6UdgjBV:bo7c32k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b05cd5f371567603d601589902128dd2c296bd5/pkg/sql/trace_test.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: no need for the semicolon. 

no placeholders here, huh?
</blockquote></details>

Fixed. I tried placeholders, I don't think they're supposed to work here:  `trace_test.go:104: pq: got 1 parameters but the statement requires 0`

---

*[pkg/util/tracing/tracer.go, line 201 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ9-IxStLMQquZhp24:-KlZdiJm1nefCsB0-plC:b-20kjiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf9ececbaa0d3d1bb9ce311ddea313a8239de2f3/pkg/util/tracing/tracer.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

subjectively I find this stuff harder to read than the previous switch-case style.
</blockquote></details>

this is not a cosmetic change. We don't want to read `enableNetTrace.Get()` and `getLightstep()` in the most common path of starting a child span from a noop span.

---

*[pkg/util/tracing/tracer.go, line 80 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ_s4hUD3J9fFNVB0A:-KlZemdzC9Wdg_PHmE6l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

optional: inline the `:=` here
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 96 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZaAXqeP1RKK47g0Ti:-KlZerAzUzzAa6DnpZaV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L96)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

golf nit: easier to read `if ptr := ...; ptr != nil { return ...}; return nil`
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 291 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZb4aqE0HOZr8uivrV:-KlZex4cpIXO1EMiEaE_:b7mb0ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/042c6f35110f28fa3dd325efff9a8c97c840af5e/pkg/util/tracing/tracer.go#L291)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the difference between `s.lightstep.Tracer()` and `lsTr`?
</blockquote></details>

Fixed (was left over from an earlier iteration).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/kv_test.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZ5SFgLrKGK2YJr9S:-KlZmG_bIPdP3E0oW3XZ:b-s41fzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/kv_test.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Tracing is now disabled by default and I doubt that will change. I kept the disablement in some tests where it was easy enough but I don't think it's necessary.
</blockquote></details>

I see. The inconsistency does seem troubling, though.

---

*[pkg/sql/trace_test.go, line 101 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZPrB2lJ2Oo0rbqg0:-KlZmY6GuM0GO370ytSw:b-nvnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b05cd5f371567603d601589902128dd2c296bd5/pkg/sql/trace_test.go#L101)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Fixed. I tried placeholders, I don't think they're supposed to work here:  `trace_test.go:104: pq: got 1 parameters but the statement requires 0`
</blockquote></details>

This update landed in the wrong commit =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 5 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/ccl/sqlccl/kv_test.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZ5SFgLrKGK2YJr9S:-KlZnlyEQ8RG-2I4DQiT:bid5lb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/kv_test.go#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I see. The inconsistency does seem troubling, though.
</blockquote></details>

I removed the others as well.

---

*[pkg/sql/trace_test.go, line 101 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZPrB2lJ2Oo0rbqg0:-KlZnQa-Md1ioQtMELyj:b9534ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b05cd5f371567603d601589902128dd2c296bd5/pkg/sql/trace_test.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

This update landed in the wrong commit =/
</blockquote></details>

Damn nits.. fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/kv_test.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZZ5SFgLrKGK2YJr9S:-KlZpGDy0Sta339cu202:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d85f26d19e5782c6d60cba064ad1feb347b96c/pkg/ccl/sqlccl/kv_test.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I removed the others as well.
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r3, 1 of 2 files at r5, 7 of 7 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/util/tracing/tracer.go, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265#-KlZ8Jwb1q-fPqIIoEZF:-KlZZpbtclQy1IessCmu:b-c8e5ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3409859c25c6f47a9411d7c3bceeaddae4c2e2/pkg/util/tracing/tracer.go#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The `len(opts) == 0` case is moved below (we need to read the options for that case).

The comment above explains it - the common cases are to have no option or just a parent reference option. This is all just to avoid doing any allocations when tracing is disabled.
</blockquote></details>

Ah, sorry, didn't read this properly. Thanks for explaining.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16265)*
<!-- Sent from Reviewable.io -->
